Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/AdsReporting.js

```diff
@@ -0,0 +1,49 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+if (webkit.messageHandlers.adsMediaReporting) {
+    install();
+}
+
+function install() {
+  function sendMessage(playing) {
+    webkit.messageHandlers.adsMediaReporting.postMessage(playing);
+  }
+
+  function checkVideoNode(node) {
+    if (node.constructor.name == "HTMLVideoElement") {
+      hookVideoFunctions();
+    }
+  }
+
+  function mediaPaused() {
+    sendMessage(false)
+  }
+
+  function mediaPlaying() {
+    sendMessage(true)
+  }
+
+  function getVideoElements() {
+    return document.querySelectorAll('video')
+  }
+
+  function hookVideoFunctions() {
+    getVideoElements().forEach(function (item) {
+      item.addEventListener('pause', mediaPaused, false);
```

Comment:
You mentioned in previous review that we should be removing event listeners before. Should we be doing this here? Since you re-query all video elements on each observer?

Reply:
Yeah since we were using anonymous functions. But if you make a named function then the duplicity is ignored by webkit itself.